### PR TITLE
Delivery Service Regional field: Optionally make Max Origin Connections per Cache Group

### PR DESCRIPTION
**Pull request by zrhoffman**
This PR resolves the Traffic Ops and Traffic Portal portions of #7097 by adding the `Regional` flag to Delivery Services for API versions 4 and up.
<hr/>

## Which Traffic Control components are affected by this PR?

- Traffic Ops

## What is the best way to verify this PR?

- Run the API tests
- Check the documentation

## PR submission checklist
- [x] This PR has tests 
- [x] This PR has documentation 
- [x] This PR has a CHANGELOG.md entry 
- [x] This PR **DOES NOT FIX A SERIOUS SECURITY VULNERABILITY** (see [the Apache Software Foundation's security guidelines](https://apache.org/security) for details)

### CODE REVIEW

File: traffic_ops/app/db/migrations/2022100210472946_ds_regional.up.sql [hunk 1]
```diff
@@ -0,0 +1,19 @@
...
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.  See the
+ * License for the specific language governing permissions and limitations under
+ * the License.
+ */
+
+ALTER TABLE public.deliveryservice 
+ADD COLUMN regional BOOLEAN NOT NULL DEFAULT FALSE;
```
[the remark this answers is not in the document]
Comment (zrhoffman) [reply traffic_ops/app/db/migrations/2022100210472946_ds_regional.up.sql:19]: oh I see what you mean now

[the remark this answers is not in the document]
Comment (zrhoffman) [reply traffic_ops/app/db/migrations/2022100210472946_ds_regional.up.sql:19]: Affecting DSRs in the migration in 7ce5da9079

**Comment by zrhoffman** [comment]
Added the Regional field to Traffic Portal v1 in 9b4c1f3f2e

File: docs/source/overview/delivery_services.rst [hunk 2]
```diff
@@ -505,6 +505,9 @@ Max Origin Connections
 ----------------------
 The maximum number of TCP connections individual :term:`Mid-tier cache servers` are allowed to make to the `Origin Server Base URL`. A value of ``0`` in this field indicates that there is no maximum.
 
+
+		.. note:: Max Origin Connections can be made per-:ref:`Cache Group <cache-groups>` by setting the :ref:`ds-regional` field.
```
Comment (ocket8888) [inline docs/source/overview/delivery_services.rst:509]: indentation here is causing this note to be rendered as a block quote - it's not a quote, so it shouldn't be indented.

File: CHANGELOG.md [hunk 3]
```diff
@@ -12,6 +12,7 @@ The format is based on [Keep a Changelog](http://keepachangelog.com/en/1.0.0/).
 - [#2101](https://github.com/apache/trafficcontrol/issues/2101) *Traffic Portal* Added the ability to tell if a Delivery Service is the target of another steering DS.
 - [#6033](https://github.com/apache/trafficcontrol/issues/6033) *Traffic Ops, Traffic Portal* Added ability to assign multiple server capabilities to a server.
 - [#7032](https://github.com/apache/trafficcontrol/issues/7032) *Cache Config* Add t3c-apply flag to use local ATS version for config generation rather than Server package Parameter, to allow managing the ATS OS package via external tools. See 'man t3c-apply' and 'man t3c-generate' for details.
+- [#7097](https://github.com/apache/trafficcontrol/issues/7097) *Traffic Ops* Added the `regional` field, which indicates whether `maxOriginConnections` should be per Cache Group
```
[the remark this answers is not in the document]
Comment (zrhoffman) [reply CHANGELOG.md]: Clarified what it was added to in f7903f022e

File: traffic_ops/app/db/migrations/2022100210472946_ds_regional.up.sql [hunk 4]
```diff
@@ -0,0 +1,30 @@
...
+
+/* Set `regional` to `false` if it does not exist */
+UPDATE public.deliveryservice_request
+SET deliveryservice = deliveryservice || '{"regional": false}'
+WHERE deliveryservice->>'regional' IS NULL;
+
+/* Set `regional` to `false` it does not exist and `original` is not null */
+UPDATE public.deliveryservice_request
+SET original = original || '{"regional": false}'
+WHERE original IS NOT NULL
```
[the remark this answers is not in the document]
Comment (zrhoffman) [reply traffic_ops/app/db/migrations/2022100210472946_ds_regional.up.sql:30]: When creating a DSR to POST a Delivery Service in Traffic Portal, the `deliveryservice` field is set but the `original` field is `NULL`. You're right that the `deliveryservice` field is nullable, too, though. Added a null check for the `deliveryservice` field in 25d28d426a

File: docs/source/overview/delivery_services.rst (hunk 2 above)
Comment (zrhoffman) [reply docs/source/overview/delivery_services.rst]: Unindented in 1640dc9ec3

File: CHANGELOG.md [hunk 5]
```diff
@@ -12,6 +12,7 @@ The format is based on [Keep a Changelog](http://keepachangelog.com/en/1.0.0/).
 - [#2101](https://github.com/apache/trafficcontrol/issues/2101) *Traffic Portal* Added the ability to tell if a Delivery Service is the target of another steering DS.
 - [#6033](https://github.com/apache/trafficcontrol/issues/6033) *Traffic Ops, Traffic Portal* Added ability to assign multiple server capabilities to a server.
 - [#7032](https://github.com/apache/trafficcontrol/issues/7032) *Cache Config* Add t3c-apply flag to use local ATS version for config generation rather than Server package Parameter, to allow managing the ATS OS package via external tools. See 'man t3c-apply' and 'man t3c-generate' for details.
+- [#7097](https://github.com/apache/trafficcontrol/issues/7097) *Traffic Ops, Traffic Portal* Added the `regional` field to Traffic Ops and Traffic Portal, which indicates whether `maxOriginConnections` should be per Cache Group
```
Comment (ocket8888) [inline CHANGELOG.md:15]: lol no I meant "to what object has a field been added?"

Delivery Services, that is.

File: traffic_ops/app/db/migrations/2022100210472946_ds_regional.up.sql (hunk 4 above)
[the remark this answers is not in the document]
Comment (zrhoffman) [reply traffic_ops/app/db/migrations/2022100210472946_ds_regional.up.sql]: Neat, removed from both in 8a099c0af0

File: traffic_portal/app/src/common/modules/form/deliveryService/form.deliveryService.DNS.tpl.html [hunk 6]
```diff
@@ -475,6 +475,26 @@ <h3 ng-if="!open()">Previous Value</h3>
...
+                            <div class="helptext">
+                                If your Delivery Service is Regional, then you only expect it to be used by a single Cache Group. Enable this to make Max Origin Connections per Cache Group instead of a value used over all last-tier Cache Groups. <a href="https://traffic-control-cdn.readthedocs.io/en/latest/overview/delivery_services.html#regional" target="_blank">See Regional Delivery Services</a>
+                            </div>
+                        </div>
+                        </label>
+                        <div class="col-md-10 col-sm-10 col-xs-12">
+                            <select id="regional" name="regional" class="form-control" ng-model="deliveryService.regional" required>
+                                <option ng-value="true">Enabled</option>
+                                <option ng-value="false" selected>Disabled</option>
+                            </select>
```
[the remark this answers is not in the document]
Comment (zrhoffman) [reply traffic_portal/app/src/common/modules/form/deliveryService/form.deliveryService.DNS.tpl.html]: Changed to a checkbox in 0aad7be6ff

File: CHANGELOG.md (hunk 5 above)
Comment (zrhoffman) [reply CHANGELOG.md]: Oh, right. Said *Delivery Service* in 8f0c2cda29

File: traffic_portal/test/integration/CommonUtils/API.ts [hunk 7]
```diff
@@ -296,7 +296,7 @@ export class API {
             const rand = () => Math.floor(Math.random()*255)+1;
             data.ipAddress = `${rand()}.${rand()}.${rand()}.${rand()}`;
         }
-        if(hasProperty(data, 'name')) {
+        if(hasProperty(data, 'name') && !hasProperty(data, 'noRandomize')) {
```
Comment (ocket8888) [inline traffic_portal/test/integration/CommonUtils/API.ts:299]: Does this need to handle if `noRandomize` is `false`?

Comment (zrhoffman) [reply traffic_portal/test/integration/CommonUtils/API.ts]: Making sure `noRandomize` is not `false` in 3eda6cba43

File: traffic_portal/test/integration/CommonUtils/API.ts [hunk 8]
```diff
@@ -296,7 +296,7 @@ export class API {
             const rand = () => Math.floor(Math.random()*255)+1;
             data.ipAddress = `${rand()}.${rand()}.${rand()}.${rand()}`;
         }
-        if(hasProperty(data, 'name') && !hasProperty(data, 'noRandomize')) {
+        if(hasProperty(data, 'name') && !(hasProperty(data, "noRandomize") && data.noRandomize === true)) {
```
Comment (ocket8888) [inline traffic_portal/test/integration/CommonUtils/API.ts:299]: nit: booleans shouldn't be compared to boolean constants, in general.

**Review by ocket8888 — APPROVED** [review]

**Pull request merged**